### ISSUE TITLE
volume mount fails: invalid mode 

### ISSUE THREAD

**Issue by rikribbers** [docker/for-win#1669]
I have an issue with docker and kubernetes on windows. I have a simple postgres setup that uses a persistence volume claim to retain its data.

When creating the pod I it does not start and ends up in a CreateContainerError status. When digging into it further I see there an invalid mode in the mount. 

"Error: Error response from daemon: invalid mode: /var/lib/postgresql/data"

### Expected behaviour

Postgres up and running

### Actual behavior

pod is in CreateContainerError  status

### Information
Below are reference to the yml and the commandline output.

Any pointer to a working setup would be appreciated.

Kubernetes yml file: 

https://gist.github.com/rikribbers/3b2c128d3a5c2599a687e562a8e865d7

Commandline output from powershell terminal

https://gist.github.com/rikribbers/10a45aac10867d83a364e5aa0b400818

docker info:
Client:
 Version:       18.02.0-ce-rc2
 API version:   1.36
 Go version:    go1.9.3
 Git commit:    f968a2c
 Built: Thu Feb  1 00:03:45 2018
 OS/Arch:       windows/amd64
 Experimental:  false
 Orchestrator:  swarm

Server:
 Engine:
  Version:      18.02.0-ce-rc2
  API version:  1.36 (minimum version 1.12)
  Go version:   go1.9.3
  Git commit:   f968a2c
  Built:        Thu Feb  1 00:10:36 2018
  OS/Arch:      linux/amd64
  Experimental: true

kubectl version 

Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T09:42:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

### Steps to reproduce the behavior

see the powershell command in the gist.

**Comment by CRuppert** [comment]
I, too am suffering from this, with latest edge. This time with the sample mysql setup provided in the k8s docs. https://kubernetes.io/docs/tasks/run-application/run-replicated-stateful-application/

Is it perhaps that there is no dynamic provisioner set up, since I noticed the following this morning: "You need to either have a dynamic PersistentVolume provisioner with a default StorageClass, or statically provision PersistentVolumes yourself to satisfy the PersistentVolumeClaims used here?"

**Comment by rikribbers** [comment]
at least I'm not alone.... updated this morning to the latest edge release still not working..

**Comment by modelt-sap** [comment]
Same thing here, trying with mysql. Works perfectly fine with docker-compose, not with Kubernetes.

**Comment by rikribbers** [comment]
Again is nice to know I'm not alone with this... however I'm not capable of solving this.. Hope this gets picked up by someone who can....

**Comment by pgayvallet** [comment]
Hello

Thanks for the feedback. Gonna look into this.

If i understand correctly, the sample mysql setup from https://kubernetes.io/docs/tasks/run-application/run-replicated-stateful-application/ is a valid reproduction scenario ?

**Comment by rikribbers** [comment]
or you can use the gist I provided

https://gist.github.com/rikribbers/3b2c128d3a5c2599a687e562a8e865d7

**Comment by pgayvallet** [comment]
@rikribbers the hostPath in the PV must refer to a path in the VM, not directly on your host. We are not able to do the same magic that we are doing on docker volumes with rewriting on the fly. The shared drive are mounted in the `/host_mnt` folder.

So,

```
 hostPath:
    path: "C:\\temp\\soccerdb-pv"
```

should be

```
 hostPath:
    path: "/host_mnt/c/temp/soccerdb-pv"
```


With this change, I got your gist running (once having removed the reference to the missing configMap in the env section)

```
Events:
  Type     Reason                 Age               From                         Message
  ----     ------                 ----              ----                         -------
  Normal   Scheduled              2m                default-scheduler            Successfully assigned soccerdb-87bdb6c9-n668t to docker-for-desktop
  Normal   SuccessfulMountVolume  2m                kubelet, docker-for-desktop  MountVolume.SetUp succeeded for volume "soccerdb-pv"
  Normal   SuccessfulMountVolume  2m                kubelet, docker-for-desktop  MountVolume.SetUp succeeded for volume "default-token-mntwv"
  Normal   Pulled                 39s (x5 over 2m)  kubelet, docker-for-desktop  Container image "postgres:10-alpine" already present on machine
  Normal   Created                39s (x5 over 2m)  kubelet, docker-for-desktop  Created container
  Normal   Started                38s (x5 over 2m)  kubelet, docker-for-desktop  Started container
```

Unfortunately, as the multiple Created/Started event tells you, this then is gonna lead you to another issue : 
```
kubectl logs po/soccerdb-XXXXX
2018-02-19 14:39:57.111 UTC [1] FATAL:  data directory "/var/lib/postgresql/data" has wrong ownership
2018-02-19 14:39:57.111 UTC [1] HINT:  The server must be started by the user that owns the data directory.
```
Which is an old issue related to the fact that we are mounting with samba, and that the owner will never be the user of the container https://forums.docker.com/t/data-directory-var-lib-postgresql-data-pgdata-has-wrong-ownership/17963/24

Two options for you then :

1/ simply `hostMount`-ing a folder in the VM ( so outside of `/host_mnt` ). Data will persist restarts, as long as you don't `clear all data` or `factory reset`, but will not be (easily) accessible from the host.

or, if having the data host-side is a necessity : 

2/ use another kind of PV to access the host (such as NFS mount, but requires you to have an nfs server host-side, and to have the latest edge version, due for this week)

I.E for option 1 : I used `/tmp/foo` as hostPath, and all is running fine : 
```
PostgreSQL init process complete; ready for start up.

2018-02-19 14:53:02.753 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2018-02-19 14:53:02.753 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2018-02-19 14:53:02.757 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2018-02-19 14:53:02.764 UTC [44] LOG:  database system was shut down at 2018-02-19 14:53:02 UTC
2018-02-19 14:53:02.765 UTC [1] LOG:  database system is ready to accept connections
```

**Comment by rikribbers** [comment]
tl;dr  quick scan seems I have options to explore, however due to time constraints probably not before next week... will look into this and report back

**Comment by savpek** [comment]
How i can configure default behavior of dynamically created persistent volumes? Because they point that C:\... folder as default (even when i disabled disk sharing from docker).

For example:
```
pvc-51c32f72-1aea-11e8-bce9-00155d67c71a
C:\Users\psa/.docker/Volumes/nginx-claim0/pvc-51c32f72-1aea-11e8-bce9-00155d67c71a
```

I got environment running after i explicitly defined volume claims for each, however that is far from ideal since i'd like to create my local environment with identical (or near identical) configurations as our test environment in cloud is. This is kind of big mismatch if i cannot figure way around it.

**Comment by rikribbers** [comment]
thx @pgayvallet  for clarifying. I understand the things you explained, I got to the same point as you with the error on directory ownership. 

As I will be using this in dev-mode only for the time being, storing it on some path of the host works for the time being. 

Again thanks for explaining and helping out.

**Comment by sinistram** [comment]
I have also meets this issue. It'll be greate to add the @pgayvallet solution to the documentation. https://kubernetes.io/docs/concepts/storage/volumes/#hostpath
